src/circuits/r1cs.rs

```diff
@@ -0,0 +1,295 @@
+    pub a: Vec<FE>,
+    pub b: Vec<FE>,
+    pub c: Vec<FE>,
+}
+#[derive(Clone, Debug, PartialEq, Eq)]
+/// R1CS represented as a vector of constraints/gates
+/// Noticing joining all the first vectors of the constraints results
+/// in the first Matrix of the R1CS
+/// joining the second vectors results on the second matrix, and so on
+///
+pub struct R1CS {
+    pub constraints: Vec<Constraint>,
+    // These are not strictly part of a R1CS
+    // But they are data of the constraint system
+    // that is needed to generate the proof
+    pub number_of_inputs: usize,
+    pub number_of_outputs: usize,
+}
+
+impl R1CS {
+    #[allow(dead_code)]
+    pub fn new(
+        constraints: Vec<Constraint>,
+        number_of_inputs: usize,
+        number_of_outputs: usize,
+    ) -> Result<Self, CreationError> {
+        let witness_size = constraints[0].a.len();
+        // Each constraints already has the correct dimensionality
+        // a check is needed before creatin the r1cs
+        // to be sure all constraints have the same dimension
+        let all_same_length = constraints
+            .iter()
+            .all(|v| v.a.len() == constraints[0].a.len());
+
+        if !all_same_length {
+            Err(CreationError::VectorsSizeMismatch)
+        } else if number_of_inputs + number_of_outputs > witness_size - 1 {
+            Err(CreationError::InputOutputTooBig)
+        } else {
+            Ok(Self {
+                constraints,
+                number_of_inputs,
+                number_of_outputs,
+            })
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn verify_solution(self, s: &[FE]) -> bool {
+        for constraint in self.constraints {
+            if !constraint.verify_solution(s) {
+                return false;
+            }
+        }
+        true
+    }
+
+    pub fn number_of_constraints(&self) -> usize {
+        self.constraints.len()
+    }
+
+    /// Returns the size of the witness
+    /// This is the constant part, plus the of inputs + intermediate values +
+    /// outputs
+    pub fn witness_size(&self) -> usize {
+        // all constraints have the same size
+        // this is enforced in the creation
+        self.constraints[0].a.len()
+    }
+}
+
+impl Constraint {
+    /// Creates a new constraint for a,b,c vectors
+    #[allow(dead_code)]
+    pub fn new(a: Vec<FE>, b: Vec<FE>, c: Vec<FE>) -> Result<Self, CreationError> {
+        if a.len() != b.len() || a.len() != c.len() || b.len() != c.len() {
+            Err(CreationError::VectorsSizeMismatch)
+        } else {
+            Ok(Self { a, b, c })
+        }
+    }
+
+    #[allow(dead_code)]
+    pub fn verify_solution(self, s: &[FE]) -> bool {
+        multiply_vectors(&self.a, s) * multiply_vectors(&self.b, s) == multiply_vectors(&self.c, s)
+    }
+}
+
+pub fn multiply_vectors(v1: &[FE], v2: &[FE]) -> FE {
+    let mut acc = FE::new(0);
+    for i in 0..v1.len() {
+        acc += v1[i] * v2[i]
+    }
+    acc
+}
+
```

Comment:
Can we name this `dot_product` or `inner_product` ?
And maybe something more compact like `v1.iter().zip(v2).map(|(x, y)| x * y).sum::<FE>();` would work?

Reply:
Renamed to inner product. As for the expression, the sum can't be done unless we define what a sum is for an iterator of FE. But we can do it with a fold like this:

v1.iter().zip(v2).map(|(x, y)| *x * *y).fold(FE::new(0), |x,y| x +y)

I'll change it